neuralmonkey/trainers/rl_trainer.py

```diff
@@ -0,0 +1,177 @@
+     (http://www.aclweb.org/anthology/P17-1138)
+     It's recommended to set subtract_baseline = True.
+    2) sample_size > 1, normalize = True, ce_smoothing = 0.0
+     Minimum Risk Training as described in 'Minimum Risk Training for Neural
+     Machine Translation' (http://www.aclweb.org/anthology/P16-1159) (Eq. 12).
+    3) sample_size > 1, normalize = False, ce_smoothing = 0.0
+     The Google 'Reinforce' objective as proposed in 'Google’s NMT System:
+     Bridging the Gap between Human and Machine Translation'
+     (https://arxiv.org/pdf/1609.08144.pdf) (Eq. 8).
+    4) sample_size > 1, normalize = False, ce_smoothing > 0.0
+     Google's 'Mixed' objective in the above paper (Eq. 9),
+     where ce_smoothing implements alpha.
+
+    Note that 'alpha' controls the sharpness of the normalized distribution,
+    while 'temperature' controls the sharpness during sampling.
+
+    :param decoder: a recurrent decoder to sample from
+    :param reward_function: any evaluator object
+    :param subtract_baseline: avg reward is subtracted from obtained reward
+    :param normalize: the probabilities of the samples are re-normalized
+    :param sample_size: number of samples to obtain feedback for
+    :param ce_smoothing: add cross-entropy loss with this coefficient to loss
+    :param alpha: determines the shape of the normalized distribution
+    :param temperature: the softmax temperature for sampling
+    :return: Objective object to be used in generic trainer
+    """
+    check_argument_types()
+
+    reference = decoder.train_inputs
+
+    def _score_with_reward_function(references: np.array,
+                                    hypotheses: np.array) -> np.array:
+        """Score (time, batch) arrays with sentence-based reward function.
+
+        Parts of the sentence after generated <pad> or </s> are ignored.
+        BPE-postprocessing is also included.
+
+        :param references: array of indices of references, shape (time, batch)
+        :param hypotheses: array of indices of hypotheses, shape (time, batch)
+        :return: an array of batch length with float rewards
+        """
+        rewards = []
+        for refs, hyps in zip(references.transpose(), hypotheses.transpose()):
+            ref_seq = []
+            hyp_seq = []
+            for r_token in refs:
+                token = decoder.vocabulary.index_to_word[r_token]
+                if token == END_TOKEN or token == PAD_TOKEN:
+                    break
+                ref_seq.append(token)
+            for h_token in hyps:
+                token = decoder.vocabulary.index_to_word[h_token]
+                if token == END_TOKEN or token == PAD_TOKEN:
+                    break
+                hyp_seq.append(token)
+            # join BPEs, split on " " to prepare list for evaluator
+            refs_tokens = " ".join(ref_seq).replace("@@ ", "").split(" ")
+            hyps_tokens = " ".join(hyp_seq).replace("@@ ", "").split(" ")
```

Comment:
It would be good to add a postprocessor parameter to the rl_objective and replace this "hardcoded" bpe postprocessing with it.

Reply:
I think we/somebody can do it themselves when we'll need to use this with a different set of pre/postprocessors.